src/testing/reporter.py

```diff
@@ -0,0 +1,122 @@
+from functools import reduce
+from pathlib import Path
+from typing import List, Union, Optional, Dict
+
+from starkware.starkware_utils.error_handling import StarkException
+
+from src.testing.cases import PassedCase, FailedCase, BrokenTest
+from src.testing.utils import TestSubject
+
+CaseResult = Union[PassedCase, FailedCase, BrokenTest]
+
+
+def print_stark_exception(exception: StarkException):
+    indented_message = "\t" + exception.message.replace("\n", "\n\t")
+    print("Error type:")
+    print(f"\t{exception.code.name}")
+    print("Error message:")
+    print(indented_message)
+    print("Error code:")
+    print(f"\t{exception.code.value}")
+
+
+class TestReporter:
+    _collected_count: Optional[int]
+    broken_tests: List[BrokenTest]
+    failed_cases: List[FailedCase]
+    passed_cases: List[PassedCase]
+    failed_tests_by_subject: Dict[Path, List[FailedCase]]
+    collected_subjects: List[TestSubject]
+    tests_root: Path
+
+    def __init__(self, tests_root: Path):
+        self.broken_tests = []
+        self.failed_cases = []
+        self.passed_cases = []
+        self.failed_tests_by_subject = {}
+        self.collected_subjects = []
+        self._collected_count = None
+        self.tests_root = tests_root
+
+    def report(self, subject: TestSubject, case_result: CaseResult):
+        symbol = None
+        if isinstance(case_result, PassedCase):
+            symbol = "."
+            self.passed_cases.append(case_result)
+        if isinstance(case_result, FailedCase):
+            symbol = "F"
+            self.failed_cases.append(case_result)
+            try:
+                self.failed_tests_by_subject[subject.test_path].append(case_result)
+            except KeyError:
+                self.failed_tests_by_subject[subject.test_path] = [case_result]
+
+        if isinstance(case_result, BrokenTest):
+            symbol = "!"
+            self.broken_tests.append(case_result)
+        assert symbol, "Unrecognised case result!"
```

Comment:
```suggestion
        assert symbol, "Unrecognized case result!"
```

Reply:
Well both are correct, really

Reply:
"Unrecognised" is British English. It's a nitpick, you can leave it if you want.